stream.go

```diff
@@ -0,0 +1,331 @@
+package psrpc
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/livekit/psrpc/internal"
+)
+
+type Stream[SendType, RecvType proto.Message] interface {
+	Channel() <-chan *Response[RecvType]
+	Send(SendType) error
+	Close(error) error
+	Err() error
+}
+
+type ServerStream[SendType, RecvType proto.Message] interface {
+	Stream[SendType, RecvType]
+	Hijack()
+}
+
+type ClientStream[SendType, RecvType proto.Message] interface {
+	Stream[SendType, RecvType]
+}
+
+type streamAdapter interface {
+	send(ctx context.Context, msg *internal.Stream) error
+	close(streamID string)
+}
+
+type clientStream struct {
+	c     *RPCClient
+	rpc   string
+	topic string
+}
+
+func (s *clientStream) send(ctx context.Context, msg *internal.Stream) (err error) {
+	if err = s.c.bus.Publish(ctx, getStreamServerChannel(s.c.serviceName, s.rpc, s.topic), msg); err != nil {
+		err = NewError(Internal, err)
+	}
+	return
+}
+
+func (s *clientStream) close(streamID string) {
+	s.c.mu.Lock()
+	delete(s.c.streamChannels, streamID)
+	s.c.mu.Unlock()
+}
+
+type serverStream[SendType, RecvType proto.Message] struct {
+	h      *streamRPCHandlerImpl[SendType, RecvType]
+	s      *RPCServer
+	nodeID string
+}
+
+func (s *serverStream[RequestType, ResponseType]) send(ctx context.Context, msg *internal.Stream) (err error) {
+	if err = s.s.bus.Publish(ctx, getStreamChannel(s.s.serviceName, s.nodeID), msg); err != nil {
+		err = NewError(Internal, err)
+	}
+	return
+}
+
+func (s *serverStream[RequestType, ResponseType]) close(streamID string) {
+	s.h.mu.Lock()
+	delete(s.h.streams, streamID)
+	s.h.mu.Unlock()
+	s.h.handling.Dec()
+}
+
+type streamHandler[SendType, RecvType proto.Message] struct {
+	*streamImpl[SendType, RecvType]
+}
+
+func (h *streamHandler[SendType, RecvType]) Recv(msg proto.Message, err error) {
+	h.recv(msg, err)
+}
+
+func (h *streamHandler[SendType, RecvType]) Send(msg proto.Message, opts ...StreamOption) error {
+	return h.send(msg)
+}
+
+func (h *streamHandler[SendType, RecvType]) Close(cause error) error {
+	return h.close(cause)
+}
+
+type streamImpl[SendType, RecvType proto.Message] struct {
+	streamOpts
+	adapter   streamAdapter
+	handler   StreamHandler
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+	recvChan  chan *Response[RecvType]
+	streamID  string
+	mu        sync.Mutex
+	recvQueue list.List
```

Comment:
why a queue instead of a channel with `opts.ChannelSize`?

Reply:
we use both. the queue is used to preserve order without blocking

Reply:
you can avoid blocking with channels, the queue is effectively ignoring channel size and can grow indefinitely
```
select {
case s.recvChan <- r:
    // channel isn't full
default:
    // log message dropped
}
```

Reply:
growing to handle bursts was intentional, i know we can drop messages... do we want to maybe add a callback or something so the program can respond to this? if we only think this happens in failure states we probably want to give the consumer a way to respond rather than logging a failure and moving on?

Reply:
maybe the sender should be notified too so they can stop sending messages?